### PR TITLE
ML & PL post

### PR DESCRIPTION
**Pull request by MikeInnes**
Our blog post on machine learning and programming.

I've added some CSS tweaks, nothing drastic but mainly making the column a bit thinner (very wide walls of text are pretty hard to read).

### CODE REVIEW

File: blog/_posts/2017-12-05-ml&pl.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+}
+x = 1; y = 2
+z = add(‘x’, ‘y’) // ‘x+y’
+eval(z) // 3
+
+x = 4
+eval(z) // 6
+{% endhighlight %}
+
+Here we are *metaprogramming* – writing code that writes code. In this case both the meta-language and the target language are the same (JavaScript) but they could just as well be different languages (as in the C preprocessor for the C language), or we can use a data structure (an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree)) instead of a string – the principle is the same. In TensorFlow, Python serves as a meta-language for writing programs in TF’s graph-based language[^ast]. If you’re not convinced, consider that TensorFlow’s graph even supports constructs like variable scoping and control flow – but rather than using Python syntax, you manipulate these constructs through an API.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-12-05-ml&pl.md]: Show some example code for the control flow. You're assuming the reader knows how ugly TF control constructs look.

File: blog/_posts/2017-12-05-ml&pl.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+There’s a snag, though. This approach relies on simplifying assumptions (ML models won’t be recursive, or need custom gradients, right?), which make it easier to apply optimisations or deploy to small devices. Unfortunately for engineers, model complexity has increased and researchers thoroughly enjoy violating these assumptions. Models now demand conditional branching (ok, easy enough to hack in), loops for recurrence (less easy but possible), and even [recursion over trees](https://arxiv.org/pdf/1503.00075.pdf) (virtually impossible to deal with). In many areas of ML, including [neural networks](https://blog.keras.io/the-future-of-deep-learning.html) and [probabilistic programming](https://eng.uber.com/pyro/), models are becoming increasingly like programs, including ones that reason about *other* programs (e.g. [program generators](https://arxiv.org/pdf/1705.03633.pdf) and [interpreters](https://arxiv.org/abs/1605.06640)). They often include non-differentiable components like Monte Carlo Tree Search, making it ever more challenging to build runtimes that are general enough while achieving top performance.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-12-05-ml&pl.md]: This can be less declarative. The "what" is there, but not the "why". Why is recursion hard? Why is tree recursion then almost impossible? Maybe a picture showing how the computational graphs grow exponentially?

File: blog/_posts/2017-12-05-ml&pl.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+*Programmable semantics* would open up new levels of flexibility, and could be provided by a feature similar to macros. This would allow features like multi-GPU training to be built on top of the core system, by specifying where the code should have pure dataflow semantics (as opposed to imperative semantics, which are more flexible but may include side-effects that are unsafe to optimise). It could also allow the kinds of program manipulation needed by probabilistic programming languages, or the [vectorisation](https://www.cs.cmu.edu/~guyb/papers/Ble90.pdf) (batching) passes usually implemented by hand in NLP models.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-12-05-ml&pl.md]: Can you show some examples of how this could be done? I don't follow you here.

File: blog/_posts/2017-12-05-ml&pl.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+As well as the PL community, ML engineers should pay close attention to the traditional Automatic Differentiation (AD) community. ML languages can take inspiration from pioneering work on [languages designed for truly first-class derivative](https://arxiv.org/pdf/1611.03416.pdf) support. Such languages can easily mix symbolic with runtime techniques (helping with the tradeoffs mentioned above), mix forward and reverse mode AD (for improved performance and memory usage), and [differentiate GPU kernels](http://mikeinnes.github.io/2017/08/24/cudanative.html) – all with no loss in performance. Most importantly, the benefits of AD and hardware support should extend to user-written libraries (say, for statistical distributions) rather than [requiring rewrites](http://pytorch.org/docs/master/distributions.html) for compatibility with the framework.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-12-05-ml&pl.md]: The JuliaDiff tools deserve a mention here.

File: blog/_posts/2017-12-05-ml&pl.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+ML research will increasingly need more powerful type systems, user-defined types and more means for extension. Gone are the days when it was enough to hard-code support for strided arrays on NVIDIA GPUs; cutting-edge techniques like [sparse machine learning](https://people.eecs.berkeley.edu/~elghaoui/Pubs/cidu2011_final.pdf), new hardware like [TPUs](https://cloud.google.com/tpu/), [Nervana](https://www.intelnervana.com/) and [FPGAs](https://www.forbes.com/sites/moorinsights/2017/08/28/microsoft-fpga-wins-versus-google-tpus-for-ai/#118733643904), and diverse deployment targets like [ARM chips](http://www.wired.co.uk/article/google-raspberry-pi-ai) or the [iPhone’s CoreML](https://developer.apple.com/documentation/coreml) chip all call for greater levels of flexibility. [Large-scale refactoring of core C++ code](https://github.com/tensorflow/tensorflow/pull/5267/files) for each new development will not scale.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-12-05-ml&pl.md]: You might want to mention the growth of mixed precision algorithms https://nickhigham.wordpress.com/2015/10/20/the-rise-of-mixed-precision-arithmetic/

File: blog/_posts/2017-12-05-ml&pl.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+Consider a world where adding new hardware support – or new kinds of data representations – could easily be accomplished by a user in high-level code, without changes to the original system. We expect ML systems to take inspiration from existing numerical computing languages, which can [already handle these tasks](https://arxiv.org/pdf/1604.03410.pdf) with ease.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-12-05-ml&pl.md]: This is a big jump. You might want to explain why the separation of actions and representation is crucial here, and then explain this example for how the array types themselves define their implementation of high level actions. If it's expanded upon, I think it's much easier to see why this strategy can be used more generally.

Comment (ChrisRackauckas) [reply blog/_posts/2017-12-05-ml&pl.md]: That would also require a statement about how high level code generally uses a small set of constructs.

File: blog/_posts/2017-12-05-ml&pl.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+[^types]: Though we note that, internally, current systems span the gamut from fully dynamic (PyTorch and its ATen backend) to unusually static (TensorFlow’s XLA and MXNet, where all dimensions are known before the graph is run).
+
+ML engineers are increasingly interested in traditional [software engineering problems](https://papers.nips.cc/paper/5656-hidden-technical-debt-in-machine-learning-systems.pdf) like the maintenance and extension of production systems. The [ML programming model](https://medium.com/@karpathy/software-2-0-a64152b37c35) makes it harder to create abstraction barriers and interfaces between components, and re-training of a model can easily break backwards compatibility. ML languages will likely be able to incorporate solutions to these problems just as regular languages do.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-12-05-ml&pl.md]: What's required to do that? Why do the languages fail right now?

File: blog/_posts/2017-12-05-ml&pl.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+A downside to any new language is that it require a new library ecosystem, as only code written for the new runtime benefits from it. For example, rather than reusing the Python ecosystem, the TensorFlow developers need to rewrite libraries for tasks like [image processing](https://www.tensorflow.org/api_guides/python/image) and [file IO](https://www.tensorflow.org/api_docs/python/tf/TextLineReader) in the graph language, throwing out the vast effort behind projects like SciPy. This may well be the only way forward, but ML practitioners should not split from the wider numerical and HPC community. An ideal ML ecosystem is an ideal numerical one, and vice versa, and collaboration between these communities will multiply everyone’s efforts.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-12-05-ml&pl.md]: Mention that this is something that Julia tools are able to do well with

File: blog/_posts/2017-12-05-ml&pl.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
...isticated, and one could imagine them adding more features and even surface syntax to become fully-fledged programming languages[^chris]. TensorFlow’s XLA started a push towards special-purpose compiler stacks that now includes [TVM](http://tvmlang.org/), [DLVM](http://dlvm.org/), [myelin](https://github.com/google/sling/tree/master/myelin), and other ongoing work. Meanwhile, projects like the [PyTorch JIT](https://github.com/pytorch/pytorch/tree/master/torch/csrc/jit), [Gluon](https://mxnet.incubator.apache.org/api/python/gluon.html) and [Tangent](https://github.com/google/tangent) are heroic efforts to make Python itself a better modelling language, in spite of enormous challenges. Having just argued that ML is a numerical programming languages problem, we in the Julia community feel that it is an excellent substrate for experimenting with these language-level issues, and will continue to push the boundaries with projects like [Knet](https://github.com/denizyuret/Knet.jl), [Flux](https://fluxml.github.io/), [Cassette](https://github.com/jrevels/Cassette.jl), [CUDAnative](https://github.com/JuliaGPU/CUDAnative.jl), [DataFlow.jl](https://github.com/MikeInnes/DataFlow.jl), and more.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-12-05-ml&pl.md]: Make the part on Julia a new paragraph (or section?) and explain why it's different than the Python approaches. Specifically: reuse of standard library, AD through the whole thing, hardware extendability through types, incorporation of language-level metaprogramming, language-level speed optimizations, etc.

File: blog/_posts/2017-12-05-ml&pl.md (hunk 1 above)
Comment (MikeInnes) [reply blog/_posts/2017-12-05-ml&pl.md]: We haven't assumed (or argued) that TF is ugly here – the point isn't to be anti-TF but just to describe how it's like a language, so I think this suffices (although I've added some links for the interested reader).

File: blog/_posts/2017-12-05-ml&pl.md (hunk 2 above)
Comment (MikeInnes) [reply blog/_posts/2017-12-05-ml&pl.md]: Well the why is simple – TF doesn't support recursion, because it made a simplifying assumption that it wouldn't be needed. Isn't that stated above?

File: blog/_posts/2017-12-05-ml&pl.md (hunk 1 above)
Comment (ChrisRackauckas) [reply blog/_posts/2017-12-05-ml&pl.md]: I think that you should put a code snippet of what it looks like so the reader can decide for themselves whether the feel is "good" or not. I think you hit home the fact that these pieces are required, but haven't given a good feel for what this actually means to code in these frameworks, which can be done by showing the custom control constructs that things like TF use. I'm not saying you should argue that TF is ugly here, just show its code and the reader will know (or make their own judgement, but I assume the reader is wise).

File: blog/_posts/2017-12-05-ml&pl.md (hunk 2 above)
Comment (ChrisRackauckas) [reply blog/_posts/2017-12-05-ml&pl.md]: Is it? I don't remember it explicitly saying that TF cannot do it, and I didn't know it couldn't. I assumed it could, but it would be slow to generate the graph. I just did a Ctrl+f and didn't find a mention of it.

File: blog/_posts/2017-12-05-ml&pl.md (hunk 5 above)
Comment (MikeInnes) [reply blog/_posts/2017-12-05-ml&pl.md]: Would be great if we can relate it back to ML, but AFAIK most folks are going to less, rather than more, precision there.

Comment (ChrisRackauckas) [reply blog/_posts/2017-12-05-ml&pl.md]: That's the purpose of mixed precision algorithms. Basically, Float16 is not stable in all parts of the algorithm. So being able to specify that the full algorithm is Float16, but compute the Gamma function using a Float32 and then use that in an activation function as a Float16 is better than requiring the whole algorithm to use Float32 all the way through just to support that one calculation.

File: blog/_posts/2017-12-05-ml&pl.md (hunk 8 above)
Comment (MikeInnes) [reply blog/_posts/2017-12-05-ml&pl.md]: Note that our goal with this is not to evangelise Julia specifically, but just to outline the trend in ML (which may have other end points, like TF becoming a full language). In future posts we'll discuss Julia's tools in more detail and link them back to these ideas explicitly.

Comment (ChrisRackauckas) [reply blog/_posts/2017-12-05-ml&pl.md]: Well usually it's nice to add to "an ideal ML ecosystem" would ..., by mentioning some recent efforts to solve the problem. The Julia tools do this, so they should be mentioned. I'm sure there's some Nim, Rust, Python, etc. tool out there that's trying to do the same that can also get a mention, but I don't know them.

File: blog/_posts/2017-12-05-ml&pl.md (hunk 2 above)
Comment (MikeInnes) [reply blog/_posts/2017-12-05-ml&pl.md]: > This approach relies on simplifying assumptions (ML models won’t be recursive, or need custom gradients, right?)

Comment (ChrisRackauckas) [reply blog/_posts/2017-12-05-ml&pl.md]: I didn't realize that statement meant TensorFlow cannot do it. I thought it meant that TF cannot optimize it well.

File: blog/_posts/2017-12-05-ml&pl.md (hunk 5 above)
Comment (MikeInnes) [reply blog/_posts/2017-12-05-ml&pl.md]: Isn't that a trivial conversion at the right point?

File: blog/_posts/2017-12-05-ml&pl.md (hunk 7 above)
Comment (MikeInnes) [reply blog/_posts/2017-12-05-ml&pl.md]: I don't think we can give a detailed analysis of this problem beyond the references (I certainly don't know what the solution is), but I've made it clear the this is an open design problem.

File: blog/_posts/2017-12-05-ml&pl.md (hunk 5 above)
Comment (ChrisRackauckas) [reply blog/_posts/2017-12-05-ml&pl.md]: That quick example, yes. But it shouldn't be difficult to come up with a more ML example where you need some vector at a higher precision otherwise you get catastrophic cancellation, but you don't want to subject everything to a higher precision (and optimally you want as much as possible down to Float16). One could even imagine a system that partitions for optimal performance automatically.

File: blog/_posts/2017-12-05-ml&pl.md (hunk 2 above)
Comment (MikeInnes) [reply blog/_posts/2017-12-05-ml&pl.md]: Would replacing "simplifying the modelling language" with something like "restricting what the language can express" above make this limitation clearer?

**Comment by MikeInnes** [comment]
Formatting changes can be reviewed in #682.

File: blog/_posts/2017-12-05-ml&pl.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+[^chris]: Google Brain’s recent hiring of Chris Lattner, a well-known language designer and compiler developer, is an interesting development on this point.
```
Comment (malmaud) [inline blog/_posts/2017-12-05-ml&pl.md]: I don't love this part since that is just speculation; we don't know what he's working on.

Comment (ViralBShah) [reply blog/_posts/2017-12-05-ml&pl.md]: Would you love it if it was labelled as speculative?

Comment (MikeInnes) [reply blog/_posts/2017-12-05-ml&pl.md]: I think the point we're trying to make here is that ML people are becoming interested in PL expertise (aligning with our argument), not that anyone might be working on a particular project. But this isn't clear and I'll try and reword it tomorrow.

Comment (ViralBShah) [reply blog/_posts/2017-12-05-ml&pl.md]: I think a rewording that makes it clear as speculative, but evidence that ML teams are beginning to take PL more seriously - all the arguments in the article, and Google hiring a high profile PL person - lends credibility to the story. It all adds up. We are not trying to speculate anything.

File: blog/_posts/2017-12-06-ml&pl.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+While machine learning does not yet have a dedicated language, several efforts are effectively creating hidden new languages underneath a Python API (like TensorFlow) while others are reusing Python as a modelling language (like PyTorch). We’d like to ask – are new ML-tailored languages required, and if so, why? More importantly, what might the ideal ML language of the future look like?
+
+## Pig Latin, and Other Hidden Languages
+
+TensorFlow (TF) and its ilk[^tf] are [already programming languages](https://dl.acm.org/citation.cfm?doid=3088525.3088527), albeit limited ones. This may seem surprising given that one uses Python to program TF. However, consider that TF requires you to write Python code to [build an expression tree](https://www.tensorflow.org/programmers_guide/graphs) in its internal language, which it then evaluates.
+
+[^tf]: We use TensorFlow for example, but could substitute other "define-before-run" frameworks like CNTK or MXNet.
```
Comment (iblis17) [inline blog/_posts/2017-12-06-ml&pl.md]: well, MXNet has both:
- declarative API
    - low level: Symbol API which operates like TF
       (https://mxnet.incubator.apache.org/api/python/symbol/symbol.html)
    - high level: Module API (https://mxnet.incubator.apache.org/api/python/module/module.html)
- imperative API
    - low level: NDArray API (https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html)
    - high level: Gluon API (https://mxnet.incubator.apache.org/api/python/gluon/gluon.html)
